cargo-cyclonedx/src/generator.rs

```diff
@@ -169,6 +171,42 @@ impl SbomGenerator {
         component
     }
 
+    fn get_purl(package: &Package) -> Result<Purl, purl::PackageError> {
+        let mut builder = purl::PurlBuilder::new(purl::PackageType::Cargo, &package.name)
+            .with_version(package.version.to_string());
+
+        if let Some(source) = &package.source {
+            if !source.is_crates_io() {
+                match source.repr.split_once('+') {
+                    // qualifier names are taken from the spec, which defines these two for all PURL types:
+                    // https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#known-qualifiers-keyvalue-pairs
+                    Some(("git", _git_path)) => {
+                        builder = builder.with_qualifier("vcs_url", source_to_vcs_url(&source))?
+                    }
+                    Some(("registry", registry_url)) => {
+                        builder =
+                            builder.with_qualifier("repository_url", urlencode(registry_url))?
+                    }
+                    Some((source, _path)) => log::error!("Unknown source kind {}", source),
+                    None => {
+                        log::error!("No '+' separator found in source field from `cargo metadata`")
+                    }
+                }
+            }
+        } else {
+            // source is None for packages from the local filesystem.
+            // The manifest path ends with a `Cargo.toml`, so the package directory is its parent
+            let package_dir = package.manifest_path.parent().unwrap();
+            // url-encode the path to the package manifest to make it a valid URL
+            let manifest_url = format!("file://{}", urlencode(package_dir.as_str()));
```

Comment:
I don't like this at all. This encodes the absolute path for workspace projects (like this).
It makes the SBOM not reproducible.

I do understand the reasoning but I would hope that there is a better way.
I'll ask on Slack.

Reply:
Summarizing from the discussion on Slack:

`pkg:cargo/cyclonedx-bom@0.4.1?download_url=file://../cyclonedx-sbom/` (a relative url instead of an absolute one) would at least solve the reproducibility issue (for PURL; there are other sources of non-determinism).

I think the only way to deal with such cases gracefully is to bypass local checkouts completely. Instead point the tool at the package URL, have it download it by itself so that it knows where it came from (default registry, custom registry, VCS URL) and include that verified source into the PURL and metadata. But that's out of scope of this PR.